core/commands/cat.go

```diff
@@ -64,14 +61,20 @@ var CatCmd = &cmds.Command{
 			max = -1
 		}
 
-		readers, length, err := cat(req.Context(), node, req.Arguments(), int64(offset), int64(max))
+		err = req.ParseBodyArgs()
```

Comment:
This dance still annoys me. `Arguments` used to handle this by default but now it doesn't. @whyrusleeping you have any idea why @keks made this change?

Reply:
It's been a while so I can't really remember, but I think it was because @Kubuxu said he wants to be able to read the body args in a streaming fashion.

Reply:
After a short conversation with @keks, we remembered its because we want to be able to stream input arguments. In which case this still kinda doesnt let us do what we want. Lets try and figure this out...

Reply:
FIRST!!!1

Reply:
I think the streaming was implement neither before nor now, but if we always parse on request arrival time we can't optionally stream. So if we did that we'd have to break the API again when that is added.
Right now, nothing uses streamed parsing (IIRC), so it can be added later.

Reply:
Hm. Ideally, I wonder if we could just make arguments a channel. However, given that this will be a large change, IMO, we should merge and keep this as a TODO (its an internal interface and not a bug).

Reply:
I'm not sure channels are the right tool here, instead I'd propose something like `func (*Request) ParseNextArg() (Argument, error)` or so. With channels it's really weird to access the parse error.

Reply:
Hm. Good point (although we could have the channel return errors). My primary goal is to make this a single API. How about:

```go
// Collect and return all arguments as a single slice. Most commands will just use this.
// Calling this again will return an error.
func (*Request) CollectArgs() ([]Argument, error)
// returns the next n (at most) arguments, or an error if parsing any fails.
func (*Request) NextArgs(n int) ([]Argument, error)
// Parse and return the next arg.
func (*Request) NextArg() (Argument, error)
```

Thoughts?